---
title: "Analysis of snRNA-seq from [Mathys, et al. Nature 2019](https://www.nature.com/articles/s41586-019-1195-2)"
subtitle: 'Data from AMP-AD synapse.org: [syn18485175](https://www.synapse.org/#!Synapse:syn18485175)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Mathys, et al. Nature 2019}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!--- 

cd /hpc/users/hoffmg01/work/eval_methods/dreamlet
ml python git
R --vanilla

system("git pull"); rmarkdown::render("mathys_Nature_2019.Rmd");


# https://hoffmg01.u.hpc.mssm.edu/eval_methods/dreamlet/mathys_Nature_2019.html

--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```

# Load packages
```{r load.packages, cache=FALSE}
# Use cache=FALSE so that package are fully loaded each time
# This ensures that forks within mclapply() have these loaded
# Othewise, mclapply() not have access to these libraries and will fail 
#   unless the libraries are manually loaded within each fork
library(SingleCellExperiment)
library(zellkonverter)
library(DelayedArray)
library(GSEABase)
library(dreamlet)
library(muscat)
library(zenith)
library(scater)
library(tidyverse)
library(BiocParallel)
library(kableExtra)
library(qvalue)

# set block size for reading DelayedMatrix in H5AD file
# The default of 1e8 to small and increasing to the block size 
# (i.e. cache size) to 1e9 decrease run time 
# with increased memory usage
setAutoBlockSize(1e9)
```

# Read data
```{r read.data}
# read h5ad
file = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Mathys_Nature_2019/Mathys_Nature_2019.h5ad"
sce = readH5AD(file, use_hdf5=TRUE)

# Set missing PMI to mean
sce$pmi[is.na(sce$pmi)] = mean(sce$pmi, na.rm=TRUE) 
```

# Tsne embedding
```{r umap}
colData(sce) %>% 
  as.data.frame() %>%
  ggplot(aes(tsne1, tsne2, color=broad.cell.type)) + 
    geom_point() + 
    theme_classic() + 
    theme(aspect.ratio=1)
```

# Sex check
Aggregate reads across cells and plot expression on the UTY from chrY
```{r sex.check}
# pseudobulk summing across all cells
pb.all <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "projid",  
    sample_id = "projid",
    verbose = TRUE)

counts = lapply(assayNames(pb.all), function(key){
  assay(pb.all, key)
  })
counts = Reduce("+", counts)

pb.all2 = SingleCellExperiment(assays=list(counts=counts), colData= colData(pb.all))

# compute sum-factors & normalize
pb.all2 <- computeLibraryFactors(pb.all2)
pb.all2 <- logNormCounts(pb.all2)

# plot Sex
data.frame(UTY = logcounts(pb.all2)["UTY",], Sex = pb.all2$msex) %>%
  ggplot(aes(Sex, UTY)) + 
    geom_point() + 
    theme_classic() + 
    theme(aspect.ratio=1)
```

# Dreamlet analysis
```{r dreamlet}
# Pseudobulk, dropping pericytes
keep = sce$broad.cell.type != "Per"
pb <- aggregateToPseudoBulk(sce[,keep],
    assay = "counts",     
    cluster_id = "broad.cell.type",  
    sample_id = "projid",
    verbose = TRUE)

# convert sequencingBatch to factor
pb$sequencingBatch = factor(pb$sequencingBatch)

form = ~ (1|sequencingBatch) + (1|msex) + (1|Dx) + educ + braaksc

# process data
res.proc = processAssays( pb, form, min.cells=5, min.count=2, min.samples=8)
```


### Show details of processing
```{r details}
details(res.proc)
```

### Show voom-style mean-variance trends
```{r voom.plot, fig.height=18, fig.width=12}
# show voom plot for each cell clusters  
plotVoom( res.proc, ncol=4)
```

# Variance partitioning analysis
```{r vp, fig.width=8, fig.height=7}
# variancePartition
vp.sex = fitVarPart( res.proc, form, BPPARAM = SnowParam(6, progressbar=TRUE))
```

```{r vp.plot}
plotVarPart(vp.sex)
```

```{r plotPercentBars, fig.height=3}
plotPercentBars(vp.sex[vp.sex$gene %in% "UTY",])
```


## Relationship between variables
Use canonical correlation analysis to show association between variables
```{r canCorPairs, fig.height=8, fig.width=7}
form =  ~ sequencingBatch + msex + educ + pmi + Dx + braaksc + dcfdx_lv + cogdx + ceradsc

C = canCorPairs(form, colData(res.proc))

plotCorrMatrix(C)
```

# Variance Partitioning
```{r vp2, fig.width=8, fig.height=7}
# run variance partitioning analysis
#  + apoe_genotype + pmi
form =  ~ (1|sequencingBatch) + (1|msex) + educ + pmi + (1|Dx) #+ braaksc + dcfdx_lv + cogdx + ceradsc
vp.lst = fitVarPart( res.proc, form, BPPARAM=SnowParam(12, progressbar=TRUE))

plotVarPart(vp.lst)
```

# Dreamlet analysis
```{r dreamlet2}
form =  ~ (1|sequencingBatch) + (1|msex) + educ + pmi + Dx
res.dl = dreamlet(res.proc, form, BPPARAM=SnowParam(12, progressbar=TRUE))
```

## Volcano plots 
```{r volcano, fig.width=8, fig.height=7}
plotVolcano(res.dl, coef='DxAD')
```

```{r resTable}

```



### Summarize differential expression 
```{r plot.pi, fig.height=12, fig.width=12}
# Summarize differential expression for each coef and assay
df = res.dl %>%
  topTable(coef='DxAD', number=Inf) %>%
    as_tibble %>% 
    group_by(assay) %>% 
    summarise( 
      nGenes = length(adj.P.Val), 
      nDE = sum(adj.P.Val < 0.05),
      pi1 = 1 - qvalue(P.Value)$pi0) 

ymax = 1.05*max(df$nGenes)
fig1 = ggplot(df, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes expressed") +
    ylab("Cell type") +
    facet_wrap(~ coef)

ymax = 1.05*max(df$nDE)
fig2 = ggplot(df, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab("Cell type") +
    facet_wrap(~ coef)

fig3 = ggplot(df, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1])) +
    ylab("Cell type") +
    facet_wrap(~ coef)

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=3, axis="tblr", align="hv")
```    



# Gene set analysis using zenith
```{r zenith1, fig.height=10}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="SYMBOL")

# Run zenith gene set analysis on result of dreamlet 
df_gs = zenith_gsa(res.dl, go.gs, coef='DxAD')
```

## Heatmap of results
```
plotZenithResults(df_gs, 5, 2)
```

# MASHR
```{r mashr}
# run mashr model to borrow information across genes and
# cell types in estimating coefficients' posterior distribution
res_mash = run_mash(res.dl, coef='DxAD')
```

```{r volcano.mash, fig.width=8, fig.height=7}
plotVolcano(res_mash)
```

```{r plotForest}
plotForest(res_mash, "OLFM3") 
```

```{r zenith.mashr, fig.height=10}
# valid values for statistic: 
# "tstatistic", "abs(tstatistic)", "logFC", "abs(logFC)"
df_gs_mash = zenith_gsa(res_mash, go.gs)

# Heatmap of results
plotZenithResults(df_gs_mash, 5, 2)
```



Figure 1c: BEX1, STMN1, BEX3, LINGO1, MTRNR2L8, DHFR
Figure 1g: APOE


library(MAST)

sce_sub = sce[1:20,sce$broad.cell.type == 'Ex']

sca <- SceToSingleCellAssay(sce_sub, check_sanity = FALSE)
# fit <- zlm(~ Dx, sca)
fit <- zlm(~ Dx + (1|projid), sca, method="glmer", ebayes=FALSE)


lrt <- suppressMessages(lrTest(fit, "Dx"))

p_val <- lrt[, "hurdle", "Pr(>Chisq)"]








