---
title: "Empirical estimates of variance inflation factors"
subtitle: 'Examine 4 dataset'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
  self_contained: false
---



<!---

cd /Users/gabrielhoffman/workspace/repos/dreamlet_analysis/delta_method
# cd /hpc/users/hoffmg01/www/dreamlet_analysis/delta_method

ml python git pandoc 
ml gcc/11.2.0
git pull 

# rm -rf empirical_vif_cache/ empirical_vif_files/

rmarkdown::render("empirical_vif.Rmd")


--->
 

```{r knitr, echo=FALSE, message=FALSE}
suppressPackageStartupMessages(library(knitr))
options(xtable.type="html")

knitr::opts_chunk$set(
  echo=TRUE,
  warning=FALSE,
  message=TRUE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,
  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)
```

```{r library, cache=FALSE}
library(SingleCellExperiment)
library(zellkonverter)
library(ggplot2)
library(dreamlet)
library(DelayedArray)
library(tidyverse)
```

# Kfoury_CancerCell_2021
```{r Kfoury}
setAutoBlockSize(1e9)

file <- "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Kfoury_CancerCell_2021/Kfoury_CancerCell_2021.h5ad"
sce <- readH5AD(file, use_hdf5=FALSE)   

sce$cells <- factor(sce$cells, sort(levels(sce$cells)))

# create pseudobulk 
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cells",  
    sample_id = "ID")

# get expressed genes for each cell type
geneList <- getExprGeneNames(pb)

# compute weights
weightsList <- pbWeights(sce,
  cluster_id = "cells", 
  sample_id = "ID",
  geneList = geneList,
  details=TRUE)

# process expression to get genes with sufficient expression
res.proc <- processAssays(pb, ~1)

# extract results
df <- lapply(names(weightsList), function(x){
	attr(weightsList[[x]], "details") 
})
df <- bind_rows(df) 
```

## plots
```{r Kfoury.plots}
ggplot(df, aes(assay, vif)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	geom_hline(yintercept=1, color="red", linetype="dashed") +
	ylab("Variance Inflation Factor") +
	coord_flip()

ggplot(df, aes(assay, sigSq.mle)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.mle") +
	coord_flip()

ggplot(df, aes(assay, sigSq.hat)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.hat") +
	coord_flip()

df %>%
	select(Gene, assay, sigSq.hat, sigSq.mle) %>%
	pivot_longer(cols=c("sigSq.mle", "sigSq.hat")) %>%
	ggplot(aes(assay, value, color=name)) +
		# geom_violin(scale="width", color=NA) +	
		geom_boxplot(width=0.1, fill="grey70") +
		theme_classic() +
		theme(aspect.ratio=1) +
		scale_y_log10() + 
		coord_flip()
```



ggplot(df, aes(sigSq.mle, sigSq.hat)) +
	geom_point() +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_x_log10() + 
	scale_y_log10() 


df2 = df %>%
	select(Gene, assay, sigSq.mle, ncell)  %>%
	filter(assay == "Erythroid") %>%
	mutate(sigSq.hat = squeezeVar(sigSq.mle + 1e-9, pmax(1, ncell - 1), robust=FALSE)$var.post)

fig =  df2 %>%
	pivot_longer(cols=c("sigSq.mle", "sigSq.hat")) %>%
	ggplot(aes(assay, value, color=name)) +
		# geom_violin(scale="width", color=NA) +	
		geom_boxplot(width=0.1, fill="grey70") +
		theme_classic() +
		theme(aspect.ratio=1) +
		scale_y_log10() + 
		coord_flip() 

ggsave(fig, file="~/www/test.png")


a = squeezeVar(df2$sigSq.hat + 1e-10, 16, robust=FALSE)$var.post


keepList = getExprGenes(pb)


# Nathan_NatImm_2021
```{r Nathan}
path = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Nathan_NatImm_2021/"
file = paste0(path, "/Nathan_NatImm_2021.h5ad")
sce = readH5AD(file, use_hdf5=FALSE)   

# get only single-cell RNA-seq
# rest is CITE-seq proteins
only_rna = grep("prot", rownames(sce), invert=TRUE)

# create variable that distinguishes cells from the same donor
# but different batches
sce$donor_batch = with(colData(sce), paste(donor, batch))
sce$batch = factor(sce$batch)
sce$season = factor(sce$season)
sce$TB_status = factor(sce$TB_status, c("CONTROL", "CASE"))

# sort cell clusters by name
sce = sce[only_rna,sce$cluster_name != "NA"]
colData(sce) = droplevels(colData(sce))
tab = colData(sce) %>%
  as_tibble %>%
  group_by(cluster_ids, cluster_name) %>%
  filter(row_number()==1) %>%
  summarise(cluster_ids, cluster_name) %>%
  arrange(as.numeric(gsub("C-", '', cluster_ids)))

sce$cluster_name = factor(sce$cluster_name, unique(tab$cluster_name))
sce$cluster_ids = factor(sce$cluster_ids, unique(tab$cluster_ids))

# create pseudobulk 
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cluster_name",  
    sample_id = "donor_batch")

# get expressed genes for each cell type
geneList = getExprGeneNames(pb)

# compute weights
weightsList <- pbWeights(sce,
  cluster_id = "cluster_name",
  sample_id = "donor_batch",
  geneList = geneList, 
  details=TRUE)

# process expression to get genes with sufficient expression
res.proc <- processAssays(pb, ~1)

# extract results
df = lapply(names(weightsList), function(x){
	attr(weightsList[[x]], "details") 
})
df = bind_rows(df) 
```

## plots
```{r Nathan.plots}
ggplot(df, aes(assay, vif)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	geom_hline(yintercept=1, color="red", linetype="dashed") +
	ylab("Variance Inflation Factor") +
	coord_flip()

ggplot(df, aes(assay, sigSq.mle)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.mle") +
	coord_flip()

ggplot(df, aes(assay, sigSq.hat)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.hat") +
	coord_flip()

df %>%
	select(Gene, assay, sigSq.hat, sigSq.mle) %>%
	pivot_longer(cols=c("sigSq.mle", "sigSq.hat")) %>%
	ggplot(aes(assay, value, color=name)) +
		# geom_violin(scale="width", color=NA) +	
		geom_boxplot(width=0.1, fill="grey70") +
		theme_classic() +
		theme(aspect.ratio=1) +
		scale_y_log10() + 
		coord_flip()
```

```{r exit, cache=FALSE}
knitr::knit_exit(1)
```

# PsychAD pr0
```{r PsychAD}
folder = "/sc/arion/projects/psychAD/NPS-AD/public_release_0/" 
file = paste0(folder, "PsychAD_r0_Dec_28_2022.h5ad")
sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE)
assayNames(sce)[1] = "counts"
sce$Dx = factor(sce$Dx_AD, c('Control','AD'))

# Create pseudo-bulk SingleCellExperiment
pb = aggregateToPseudoBulk(sce,
    assay = "counts", 
    cluster_id = "subclass",
    sample_id = "Channel",
    BPPARAM = SnowParam(6))

# get expressed genes for each cell type
geneList = getExprGeneNames(pb)

# compute weights
weightsList <- pbWeights(sce,
  cluster_id = "subclass", 
  sample_id = "Channel",
  geneList = geneList,
  details=TRUE)

# process expression to get genes with sufficient expression
res.proc <- processAssays(pb, ~1)

# extract results
df = lapply(names(weightsList), function(x){
	attr(weightsList[[x]], "details")
})
df = bind_rows(df) 
```


## plots
```{r PsychAD.plots}
ggplot(df, aes(assay, vif)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	geom_hline(yintercept=1, color="red", linetype="dashed") +
	ylab("Variance Inflation Factor") +
	coord_flip()

ggplot(df, aes(assay, sigSq.mle)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.mle") +
	coord_flip()

ggplot(df, aes(assay, sigSq.hat)) +
	geom_violin(scale="width") +	
	geom_boxplot(width=0.1, fill="grey70") +
	theme_classic() +
	theme(aspect.ratio=1) +
	scale_y_log10() + 
	ylab("sigSq.hat") +
	coord_flip()

df %>%
	select(Gene, assay, sigSq.hat, sigSq.mle) %>%
	pivot_longer(cols=c("sigSq.mle", "sigSq.hat")) %>%
	ggplot(aes(assay, value, color=name)) +
		# geom_violin(scale="width", color=NA) +	
		geom_boxplot(width=0.1, fill="grey70") +
		theme_classic() +
		theme(aspect.ratio=1) +
		scale_y_log10() + 
		coord_flip()
```


















