---
title: "Human and mouse single-nucleus transcriptomics reveal TREM2-dependent and TREM2-independent cellular responses in Alzheimerâ€™s disease"
subtitle: '[Zhou, et al. Nature Medicine (2020)](https://www.nature.com/articles/s41591-019-0695-9)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
    toc: true
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r load.packages, cache=FALSE}
library(SingleCellExperiment)
library(zellkonverter)
library(GSEABase)
library(dreamlet)
library(scater)
library(zenith)
library(knitr)
library(kableExtra)
library(scattermore)
library(cowplot)
library(ggplot2)
library(qvalue)
library(tidyverse)
library(BiocParallel)
library(DelayedArray)

# set block size for reading DelayedMatrix in H5AD file
# The default of 1e8 to small and increasing to the block size 
# (i.e. cache size) to 1e9 decrease run time 
# with increased memory usage
setAutoBlockSize(1e9)
```

```{r load}
file = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Zhou_Nat_Met_2020/Zhou_Nat_Met_2020.h5ad"

sce = readH5AD(file, use_hdf5=TRUE) 
sce$sequencingBatch = factor(sce$sequencingBatch) 

# create pseudobulk 
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "Label",  
    sample_id = "specimenID")
```

# tSNE embedding
```{r embeding}
colData(sce) %>%
  data.frame() %>%
  ggplot(aes(tSNE_1, tSNE_2, color=Label)) + 
  geom_scattermore() +
  theme_classic() + 
  theme(aspect.ratio = 1, 
    axis.text = element_blank(),
    axis.ticks = element_blank()) +
  scale_color_discrete(name="Cell type") +
  xlab("tSNE 1") +
  ylab("tSNE 2")
```




# Sex check
Aggregate reads across cells and plot expression on the UTY from chrY to confirm proper sex assignment.
```{r sex.check}
# pseudobulk summing across all cells
pb.all <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "specimenID",  
    sample_id = "specimenID",
    verbose = TRUE)

counts = lapply(assayNames(pb.all), function(key){
  assay(pb.all, key)
  })
counts = Reduce("+", counts)

pb.all2 = SingleCellExperiment(assays=list(counts=counts), 
          colData= colData(pb.all))

# compute sum-factors & normalize
pb.all2 <- computeLibraryFactors(pb.all2)
pb.all2 <- logNormCounts(pb.all2)

df  = data.frame(Sex = ifelse(logcounts(pb.all2)["ENSG00000183878",] > 1, "M", "F"))

i = match(colnames(pb), rownames(df))
pb$Sex = df$Sex
```



## Process pseudobulk data to estimate precision weights
```{r voom}
# Normalize and apply voom/voomWithDreamWeights
form = ~ (1|Dx) + (1|sequencingBatch) + (1|Sex)
res.proc = processAssays( pb,  
                          form,  
                          min.count=5,
                          BPPARAM = SnowParam(6))
```

### Show details of processing
```{r details}
details(res.proc)
```

### Show voom-style mean-variance trends
```{r voom.plot, fig.height=18, fig.width=12}
# show voom plot for each cell clusters  
plotVoom( res.proc, ncol=4)
```

# Variance partitioning analysis
```{r vp}
# run variance partitioning analysis
form = ~ (1|Dx) + (1|sequencingBatch) + (1|Sex)
vp.lst = fitVarPart( res.proc, form, 
  BPPARAM = SnowParam(6, progressbar=TRUE))
```

```{r vp.plot, fig.height=20, fig.width=12}
# Summarize variance fractions genome-wide for each cell type
plotVarPart(sortCols(vp.lst), label.angle=60, ncol=4) 
```

# need to Model Sex



# dreamlet analysis
```{r dreamlet}
form = ~ 0 + Dx + (1|sequencingBatch)  + (1|Sex)

contrasts = c( "P_vs_C" = 'DxP - DxC',
               "AD_vs_C" = 'DxAD - DxC',
               "P_vs_AD" = 'DxP - DxAD')
# dreamlet analysis
res.dl = dreamlet(res.proc, 
                  form,                   
                  contrasts=contrasts, 
                  BPPARAM = SnowParam(6, progressbar=TRUE))
```

# Volcano plots for each contrast
### P_vs_C
```{r volcano.P_vs_C, fig.height=16, fig.width=10}
plotVolcano( res.dl, coef = 'P_vs_C', ncol=4)
```

### AD_vs_C
```{r volcano.AD_vs_C, fig.height=16, fig.width=10}
plotVolcano( res.dl, coef = 'AD_vs_C', ncol=4)
```

### P_vs_AD
```{r volcano.P_vs_AD, fig.height=16, fig.width=10}
plotVolcano( res.dl, coef = 'P_vs_AD', ncol=4)
```



### Summarize differential expression 
```{r plot.pi, fig.height=12, fig.width=12}
# Summarize differential expression for each coef and assay
df = names(contrasts) %>%
  map_df(function(x) 
    topTable(res.dl, coef = x, number=Inf) %>%
    as_tibble %>%
    mutate(coef = x)) %>%
  group_by(coef,assay) %>%
  summarize( nDE = sum(adj.P.Val < 0.05), 
            pi1 = 1 - pi0est(P.Value)$pi0,
            nGenes = length(adj.P.Val)) %>%
  mutate(assay = factor(assay, assayNames(pb)))

ymax = 1.05*max(df$nGenes)
fig1 = ggplot(df, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes expressed") +
    ylab("Cell type") +
    facet_wrap(~ coef)

ymax = 1.05*max(df$nDE)
fig2 = ggplot(df, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab("Cell type") +
    facet_wrap(~ coef)

fig3 = ggplot(df, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1])) +
    ylab("Cell type") +
    facet_wrap(~ coef)

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=3, axis="tblr", align="hv")
```    


# Gene set analysis using zenith
```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="ENSEMBL")

res_zenith_list = map(names(contrasts), function(coef) 
  zenith_gsa(res.dl, go.gs, coef))
names(res_zenith_list) = names(contrasts)
```

### P_vs_C
```{r zenith.plot_Tumor_Involved, fig.height=18, fig.width=12}
plotZenithResults(res_zenith_list[["P_vs_C"]], 5, 3)
```

### AD_vs_C
```{r zenith.plot_Tumor_Distal, fig.height=18, fig.width=12}
plotZenithResults(res_zenith_list[["AD_vs_C"]], 5, 3)
```

### P_vs_AD
```{r zenith.plot_Involved_Distal, fig.height=22, fig.width=13}
plotZenithResults(res_zenith_list[["P_vs_AD"]], 5, 3)
```







