---
title: "Human and mouse single-nucleus transcriptomics reveal TREM2-dependent and TREM2-independent cellular responses in Alzheimerâ€™s disease"
subtitle: '[Zhou, et al. Nature Medicine (2020)](https://www.nature.com/articles/s41591-019-0695-9)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
    toc: true
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```

```{r load, cache=FALSE}
library(Matrix)
library(SingleCellExperiment)
library(zellkonverter)
library(openxlsx)
```

```{r download}
path = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Zhou_Nat_Met_2020/"
setwd(path) 
system("synapse get -r syn21682218")
system("synapse get -r syn21126462")
system("synapse get -r syn22264145")
system("mkdir data; mv ./* data")
```


```{r read.data}
file = paste0(path, '/data/snRNAseqAD_TREM2_assay_scRNAseq_metadata.csv')
df_meta = read.table(file, header=TRUE, sep=',')
rownames(df_meta) = df_meta$specimenID
df_meta$Dx = gsub("(\\d+)$", "", df_meta$specimenID)
df_meta$Dx = factor(df_meta$Dx, c("C", "AD", "P"))

resList = lapply(rownames(df_meta), function(id){

	# read counts file
	file = paste0(path, '/data/', id, '_matrix.mtx.gz')
	counts = readMM(file)

	# read barcodes to columns
	file = paste0(path, '/data/', id, '_barcodes.tsv.gz')
	cn = scan(file, what=character())
	colnames(counts) = paste0(id, '_', cn)

	# read gene names
	file = paste0(path, '/data/', id, '_features.tsv.gz')
	df = read.table(file)
	rownames(counts) = df$V1

	# create colData for each cell from metadata for each sample
	info = df_meta[rep(id, ncol(counts)),]
	rownames(info) = colnames(counts)

	list(counts=counts, info = info)
	})

# combine results from each sample
counts = do.call(cbind, lapply(resList, function(x) x$counts))
info = do.call(rbind, lapply(resList, function(x) x$info))
```

```{r create.sce}
# Read cell-level metadata
file = paste0(path, "data/clusters_cellID.xlsx")
df_annotations = read.xlsx(file, sheet=2)

# Update names
df_annotations$Sample = gsub("^TWCC-", "", df_annotations$Sample)
df_annotations$Barcodes = gsub("^TWCC-", "", df_annotations$Barcodes)

# remove -1 suffix from cell identifiers
rownames(info) = gsub("-1$", "", rownames(info))
colnames(counts) = gsub("-1$", "", colnames(counts))

# Merge two sources of metadata
info2 = merge(info, df_annotations, by.x = "row.names", by.y = "Barcodes")
rownames(info2) = info2$Row.names
info2$Row.names = c()

# Create SingleCellExperiment
# Only keep cells assigned to clusters
sce = SingleCellExperiment( assays = 
		list(counts = counts[,rownames(info2)]),
		colData = info2)
```

```{r write}
# write to file 
file = paste0(path, '/Zhou_Nat_Met_2020.h5ad')
writeH5AD(sce, file, compression="lzf")
```























