---
title: "Analysis of Alzheimer's Disease in [PsychAD](https://adknowledgeportal.synapse.org/Explore/Projects/DetailsPage?Grant%20Number=R01AG067025)"
subtitle: 'Public Release 0'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```

# Load libraries
```{r load.packages, cache=FALSE}
suppressPackageStartupMessages({
library(SingleCellExperiment)
library(zellkonverter)
library(dreamlet)
library(zenith)
library(DelayedArray)

library(GSEABase)
library(muscat)
library(cowplot)
library(scater)

library(tidyverse)
library(kableExtra)
library(qvalue)
library(scattermore)
library(corrplot)
library(org.Hs.eg.db)
})
```


```{r load.data}
file = "/sc/arion/projects/psychAD/NPS-AD/public_release_0/PsychAD_r0_Oct_06_2022.h5ad"
sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE)
assayNames(sce)[1] = "counts"

# drop cells with annotation "Unknown"
sce = sce[,sce$celltype != "Unknown"]
colData(sce) = droplevels(colData(sce)) 

# AD status as factor
sce$AD = factor(sce$AD, 0:1)
```

Public freeze 0 includes `r length(table(sce$Channel))` samples, `r length(table(sce$round_num))` rounds, `r length(table(sce$batch))` 10X batches, `r length(table(sce$SubID))` donors, and `r format(ncol(sce), big.mark=',')` cells passing QC.

# Joint UMAP
```{r umap, dev="png"}
# extract UMAP coordinates and annotations
df = cbind(reducedDim(sce, "X_umap"), 
    colData(sce)[,c("anno", "celltype", "class", "subtype")]) %>% 
    as.data.frame
 
ggplot(df, aes(V1, V2, color=celltype)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2")
```

```{r umap3, dev="png"}
ggplot(df, aes(V1, V2, color=class)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```


## Summarize cell counts
```{r summarize.cell.counts}
# cells per Channel
colData(sce)$Channel %>% 
  table %>% 
  hist(main=paste0("Cell counts per Channel: mean=", format(mean(.), digits=2), ", median = ", format(median(.), digits=2)))

colData(sce)$SubID %>% 
  table %>% 
  hist(main=paste0("Cell counts per SubID: mean=", format(mean(.), digits=2), ", median = ", format(median(.), digits=2)))

# Number of cells observed per Channel
colData(sce) %>%
  xtabs( ~ Channel + celltype,.) %>%
  as_tibble %>%
  pivot_longer(cols=Channel) %>%
  ggplot(aes(celltype, n, fill=celltype)) + 
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of cells observed per Channel") +
    xlab('')

# Number of cells observed per Subject
colData(sce) %>%
  xtabs( ~ SubID + celltype,.) %>%
  as_tibble %>%
  pivot_longer(cols=SubID) %>%
  ggplot(aes(celltype, n, fill=celltype)) + 
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of cells observed per Subject") +
    xlab('')
```


```{r combineData, message=TRUE}
# update block size for reading h5ad file from disk
setAutoBlockSize(1e9)

# Create pseudo-bulk SingleCellExperiment
pb = aggregateToPseudoBulk(sce,
    assay = "counts", 
    cluster_id = "celltype",
    sample_id = "Channel",
    BPPARAM = SnowParam(6))
```


## Summarize read counts
```{r summarize.read.counts}
# extract read counts for each Channel
df_counts = lapply( assayNames(pb), function(x){
  data = assay(pb, x)

  data.frame(celltype = x, ID = colnames(data), readCounts = colSums(data))

})
df_counts = do.call(rbind, df_counts)

ggplot(df_counts, aes(celltype, readCounts, fill=celltype)) +  
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of reads observed for each Channel") +
    xlab('') +
    ggtitle('Reads per cell cluster for each Channel')

# extract cell counts
df_rate = cellCounts(pb) %>%
            as.data.frame %>%
            mutate(ID = rownames(.))  %>% 
            pivot_longer(cols=-ID, values_to="ncells", names_to="celltype")

# plot reads per cell
inner_join(df_counts, df_rate, by=c("celltype", "ID")) %>%
  ggplot(aes(celltype, readCounts/ncells, fill=celltype)) +  
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab('Reads per cell') +
    xlab('') +
    ggtitle('Reads per cell for each Channel')
```

## examine total reads and reads per cluster





## Cell type specificity
```{r cellTypeSpecificity}
df = cellTypeSpecificity( pb )

plotViolin(df)
```

### Show cell markers
```{r cellMarkers}
library(org.Hs.eg.db)
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_brain_cell_markers.json
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_brain_immune_cell_markers.json
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_immune_cell_markers.json
genes = c('RBFOX3', 'MEG3', 'SLC17A7',
'RBFOX3', 'MEG3', 'GAD1', 'GAD2', 'GRIK1',
'SST',
'PVALB',
'VIP',
'SLC1A3', 'GFAP', 'APOE', 'SLC1A2', 'SLC14A1',
'PLP1', 'MAG', 'MBP',
'PDGFRA', 'VCAN',
'FLT1', 'CLDN5',
'PDGFRB',
'TGFBR1', 'DOCK8', 'CD74', 'CSF1R', 'MS4A6A', 'PLXDC2')
 
df_genes = AnnotationDbi::select(org.Hs.eg.db, genes, "ENSEMBL", "SYMBOL")

df_sub = df[rownames(df) %in% df_genes$ENSEMBL,]
idx = match(rownames(df_sub), df_genes$ENSEMBL)
rownames(df_sub) = df_genes$SYMBOL[idx]

plotPercentBars(df, genes=unique(genes))
```

```{r heatmap, fig.height=7, fig.width=7}
dreamlet::plotHeatmap(df, genes=unique(genes))
```




```{r Dx.summary, eval=FALSE}
df = unique(data.frame(Dx = pb$dx, Donor = pb$SubID))

sort(table(df$dx), decreasing=TRUE) %>% kbl() %>% kable_styling(full_width=FALSE)
```

```{r voom}
# Normalize and apply voom
form = ~ 1
res.proc = processAssays( pb, form, 
  BPPARAM = SnowParam(6))
  
res.proc
```

```{r voom.plot, fig.height=10, fig.width=10, cache=TRUE}
plotVoom( res.proc ) 
```


```{r plot.sex} 
df_gene =  AnnotationDbi::select(org.Hs.eg.db, keys=c("XIST", "UTY"), keytype="SYMBOL", columns="ENSEMBL")

df_sex = lapply( names(res.proc), function(CT){
  geneExpr = res.proc[[CT]]
  df = NULL
  # if( sum(df_gene$ENSEMBL %in% rownames(geneExpr)) == 2){  
  #   df = data.frame(cellType = CT, IDS = colnames(geneExpr$E), XIST = geneExpr$E[df_gene$ENSEMBL[1],], UTY = geneExpr$E[df_gene$ENSEMBL[2],])
  # }
  if( sum(df_gene$SYMBOL %in% rownames(geneExpr)) == 2){
    df = data.frame(cellType = CT, IDS = colnames(geneExpr$E), XIST = geneExpr$E['XIST',], UTY = geneExpr$E['UTY',])
  }
  df
}) 
df_sex = do.call(rbind, df_sex)

df_sex = merge(df_sex, colData(pb)[,"Sex",drop=FALSE], by.x="IDS", by.y="row.names")

ggplot(as.data.frame(df_sex), aes(XIST, UTY, color=Sex)) + geom_point() + theme_classic() + facet_wrap(~cellType) + theme(aspect.ratio=1)
```

Sex expression based on XIST and UTY
```{r sex.combine, eval=FALSE}
# get total reads
totalReads = DelayedMatrixStats::colSums2(assay(sceCombine, "X"))
df_totalReads = data.frame(totalReads, Channel=sceCombine$Channel) %>%
  group_by(Channel) %>%
  summarise(LibSize = sum(totalReads))

rm(totalReads) 
gc()

# sum across cells for each Donor
counts = assay(sceCombine[df_gene$ENSEMBL,], "X")

setAutoBlockSize(1e8)
grid = colAutoGrid(counts, ncol=100000)

df = dreamlet:::colsum_fast(counts, droplevels(sceCombine$Channel), grid=grid)

# merge expression with metadata
df_meta_uniq = unique(colData(sceCombine)[,c("Channel", "Sex")])
df_meta_uniq = merge(df_meta_uniq, df_totalReads, by="Channel")

df_sex = merge(df_meta_uniq, as.matrix(t(df)), by.x="Channel", by.y="row.names")

rm(df, df_meta_uniq)
gc()

colnames(df_sex)[colnames(df_sex) == df_gene$ENSEMBL[1]] = df_gene$SYMBOL[1] 
colnames(df_sex)[colnames(df_sex) == df_gene$ENSEMBL[2]] = df_gene$SYMBOL[2]

df_sex2 = df_sex %>% 
          group_by(Channel) %>%
          summarise(Sex = unique(Sex),
                    UTY = log2(sum(UTY) + 0.25) - log2(sum(LibSize)) + log2(1e6),
                    XIST= log2(sum(XIST) + 0.25) - log2(sum(LibSize)) + log2(1e6))

ggplot(df_sex2, aes(XIST, UTY, color=Sex)) + geom_point() + theme_classic() + scale_color_manual(values=c("red", "blue", "green"))
```





```{r varPart}
form = ~ (1|SubID) + (1|batch) + (1|Sex) + scale(Age) + (1|AD)

res.vp = fitVarPart(res.proc, form, BPPARAM = SnowParam(6))
```

```{r vp.plot, fig.height=15, fig.width=10}
colnames(res.vp)[colnames(res.vp) == "scale.Age."] = 'Age'
colnames(res.vp)[colnames(res.vp) == "batch"] = 'Batch'
colnames(res.vp)[colnames(res.vp) == "SubID"] = 'Subject'

plotVarPart( sortCols(res.vp), label.angle=45 )
```

```{r plotPercentBars, fig.height=6, fig.width=10}
# Show variance fractions at the gene-level for each cell type
genes = c("KCNMA1", 'RASGRF2', "CECR2", "PTPRG", "XIST", 'C1D')

# res.vp %>%
#  as_tibble %>%
#  filter(assay=="Microglia") %>%
#  summarize(Age = gene[which.max(Age)],
#    Subject = gene[which.max(Subject)],
#    Batch = gene[which.max(Batch)],
#    Residuals = gene[which.max(Residuals)])

# res.vp %>%
#  as_tibble %>%
#  filter(assay=="Microglia") %>%
#  summarize(gene[order(Batch, decreasing=TRUE)][1:5])


df.vp.sub = sortCols(res.vp)[(res.vp$gene %in% genes) &(res.vp$assay=="Microglia"),]
df.vp.sub$gene = factor(df.vp.sub$gene, rev(genes))
df.vp.sub = df.vp.sub[order(df.vp.sub$gene, decreasing=TRUE),]

plotPercentBars(df.vp.sub)
```

# Correlation with and between donors
```{r within_btw, fig.height=10, fig.width=10}
source("/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/common_analyses.R")

eval_within_across_donor( res.proc, "SubID" )
```

# dreamlet
```{r dreamlet}
form = ~ (1|SubID) + (1|batch) + (1|Sex) + scale(Age) + AD

fit = dreamlet( res.proc, form, BPPARAM = SnowParam(6))
```

## Volcano
```{r volcano, fig.height=15, fig.width=10}
plotVolcano(fit, "AD1")
```

```{r genehatmap, fig.height=4, fig.width=6}
plotGeneHeatmap( fit, coef="AD1", genes=genes)
```

```{r plotForest, fig.height=6, fig.width=6}
plotForest( fit, coef = 'AD1', gene = genes[1]) +
  theme(aspect.ratio=1)
```


### Summarize differential expression 
```{r plot.pi, fig.height=5, fig.width=12}
# Summarize differential expression for each coef and assay
df = fit %>%
  topTable(coef='AD1', number=Inf) %>%
    as_tibble %>% 
    group_by(assay) %>% 
    summarise( 
      nGenes = length(adj.P.Val), 
      nDE = sum(adj.P.Val < 0.05),
      pi1 = 1 - qvalue(P.Value)$pi0) 

ymax = 1.05*max(df$nGenes)
fig1 = ggplot(df, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes expressed") +
    ylab("Cell type") 

ymax = max(1.05*max(df$nDE), 100)
fig2 = ggplot(df, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none", axis.text.y=element_blank()) +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab('')

fig3 = ggplot(df, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none", axis.text.y=element_blank()) +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1]))+
    ylab('')

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=1, axis="tblr", align="hv")
```    



## Gene set analysis
```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="SYMBOL")

# zenith analysis
res.gsa = zenith_gsa(fit, go.gs, coefs="AD1", n_genes_min=30)
```

```{r zenith.heatmap.1, fig.width=10, fig.height=18, cache=FALSE} 
plotZenithResults(res.gsa, 5, 2) + theme(legend.position="bottom") 
```

# Write residuals



# Session Info
<details>
```{r sessioninfo, cache=FALSE}
sessionInfo()
```
</details>



```{r exit2, cache=FALSE}
knitr::knit_exit()
```












