---
title: "Analysis of Alzheimer's Disease in [PsychAD](https://adknowledgeportal.synapse.org/Explore/Projects/DetailsPage?Grant%20Number=R01AG067025)"
subtitle: 'Public Release 0'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```

# Load libraries
```{r load.packages, cache=FALSE}
suppressPackageStartupMessages({
library(SingleCellExperiment)
library(zellkonverter)
library(dreamlet)
library(zenith)
library(DelayedArray)

library(GSEABase)
library(muscat)
library(cowplot)
library(scater)

library(tidyverse)
library(kableExtra)
library(qvalue)
library(scattermore)
library(corrplot)
library(dplyr)
library(org.Hs.eg.db)
library(EnrichmentBrowser)
library(mashr)
})
```


```{r load.data}
folder = "/sc/arion/projects/psychAD/NPS-AD/public_release_0/" 
file = paste0(folder, "PsychAD_r0_Oct_06_2022.h5ad")
sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE)
assayNames(sce)[1] = "counts"

# drop cells with annotation "Unknown"
sce = sce[,sce$celltype != "Unknown"]
colData(sce) = droplevels(colData(sce)) 

# AD status as factor
sce$AD = factor(sce$AD, 0:1)

# collapse cell type annotations into 9 categories
sce$celltype9 = recode_factor(sce$subtype,
                Astro_1 = "Astrocyte",
                Astro_2 = "Astrocyte",
                EN_L2_3_IT_1 = "Excitatory neuron",
                EN_L2_3_IT_2 = "Excitatory neuron",
                EN_L4_IT = "Excitatory neuron",
                EN_L5_6_NP = "Excitatory neuron",
                EN_L5_IT = "Excitatory neuron",
                EN_L6_CT = "Excitatory neuron",
                Endo = "Endothelial",
                IN_Chandelier = "Inhibitory neuron",
                IN_LAMP5 = "Inhibitory neuron",
                IN_PVALB = "Inhibitory neuron",
                IN_SST = "Inhibitory neuron",
                IN_VIP = "Inhibitory neuron",
                Immune_T = "T cell",
                Micro_PVM_1 = "Microglia",
                Micro_PVM_2 = "Microglia",
                OPC = "OPC",
                Oligo_1 = "Oligodendrocyte",
                Oligo_2 = "Oligodendrocyte",
                Oligo_3 = "Oligodendrocyte",
                Oligo_4 = "Oligodendrocyte",
                Oligo_5 = "Oligodendrocyte",
                Oligo_6 = "Oligodendrocyte",
                Oligo_7 = "Oligodendrocyte",
                VLMC_SMC_PC = "VLMC_SMC_PC")

```

Public freeze 0 includes `r length(table(sce$Channel))` samples, `r length(table(sce$round_num))` rounds, `r length(table(sce$batch))` 10X batches, `r length(table(sce$SubID))` donors, and `r format(ncol(sce), big.mark=',')` cells passing QC.

# Joint UMAP
```{r umap, dev="png"}
# extract UMAP coordinates and annotations
df = cbind(reducedDim(sce, "X_umap"), 
    colData(sce)[,c("anno", "celltype", "celltype9", "class", "subtype")]) %>% 
    as.data.frame
 
ggplot(df, aes(V1, V2, color=celltype)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2")
```

```{r umap3, dev="png"}
ggplot(df, aes(V1, V2, color=class)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```


```{r umap4, dev="png"}
ggplot(df, aes(V1, V2, color=anno)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```

```{r umap9, dev="png"}
ggplot(df, aes(V1, V2, color=celltype9)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```



```{r umap5, dev="png"}
ggplot(df, aes(V1, V2, color=subtype)) + geom_scattermore() + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```


## Summarize cell counts
```{r summarize.cell.counts}
# cells per Channel
colData(sce)$Channel %>% 
  table %>% 
  hist(main=paste0("Cell counts per Channel: mean=", format(mean(.), digits=2), ", median = ", format(median(.), digits=2)))

colData(sce)$SubID %>% 
  table %>% 
  hist(main=paste0("Cell counts per SubID: mean=", format(mean(.), digits=2), ", median = ", format(median(.), digits=2)))

# Number of cells observed per Channel
colData(sce) %>%
  xtabs( ~ Channel + celltype9,.) %>%
  as_tibble %>%
  pivot_longer(cols=Channel) %>%
  ggplot(aes(celltype9, n, fill=celltype9)) + 
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of cells observed per Channel") +
    xlab('')

# Number of cells observed per Subject
colData(sce) %>%
  xtabs( ~ SubID + celltype9,.) %>%
  as_tibble %>%
  pivot_longer(cols=SubID) %>%
  ggplot(aes(celltype9, n, fill=celltype9)) + 
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of cells observed per Subject") +
    xlab('')
```


```{r combineData, message=TRUE}
# update block size for reading h5ad file from disk
setAutoBlockSize(1e9)

# Create pseudo-bulk SingleCellExperiment
pb = aggregateToPseudoBulk(sce,
    assay = "counts", 
    cluster_id = "celltype9",
    sample_id = "Channel",
    BPPARAM = SnowParam(6))
```


## Summarize read counts
```{r summarize.read.counts}
# extract read counts for each Channel
df_counts = lapply( assayNames(pb), function(x){
  data = assay(pb, x)

  data.frame(celltype = x, ID = colnames(data), readCounts = colSums(data))

})
df_counts = do.call(rbind, df_counts)

ggplot(df_counts, aes(celltype, readCounts, fill=celltype)) +  
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab("Number of reads observed for each Channel") +
    xlab('') +
    ggtitle('Reads per cell cluster for each Channel')

# extract cell counts
df_rate = cellCounts(pb) %>%
            as.data.frame %>%
            mutate(ID = rownames(.))  %>% 
            pivot_longer(cols=-ID, values_to="ncells", names_to="celltype")

# plot reads per cell
inner_join(df_counts, df_rate, by=c("celltype", "ID")) %>%
  ggplot(aes(celltype, readCounts/ncells, fill=celltype)) +  
    geom_violin(color = NA) + 
    geom_boxplot(width=.1, outlier.size=.1) +
    theme_classic() + 
    theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5),
      legend.position="none") + 
    scale_y_log10() +
    coord_flip() +
    ylab('Reads per cell') +
    xlab('') +
    ggtitle('Reads per cell for each Channel')
```

## examine total reads and reads per cluster





## Cell type specificity
```{r cellTypeSpecificity}
df = cellTypeSpecificity( pb )

plotViolin(df)
```

### Show cell markers
```{r cellMarkers}
library(org.Hs.eg.db)
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_brain_cell_markers.json
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_brain_immune_cell_markers.json
# /sc/arion/projects/CommonMind/leed62/app/leetools/pegasus/human_immune_cell_markers.json
genes = c('RBFOX3', 'MEG3', 'SLC17A7',
'RBFOX3', 'MEG3', 'GAD1', 'GAD2', 'GRIK1',
'SST',
'PVALB',
'VIP',
'SLC1A3', 'GFAP', 'APOE', 'SLC1A2', 'SLC14A1',
'PLP1', 'MAG', 'MBP',
'PDGFRA', 'VCAN',
'FLT1', 'CLDN5',
'PDGFRB',
'TGFBR1', 'DOCK8', 'CD74', 'CSF1R', 'MS4A6A', 'PLXDC2')
 
df_genes = AnnotationDbi::select(org.Hs.eg.db, genes, "ENSEMBL", "SYMBOL")

df_sub = df[rownames(df) %in% df_genes$ENSEMBL,]
idx = match(rownames(df_sub), df_genes$ENSEMBL)
rownames(df_sub) = df_genes$SYMBOL[idx]

plotPercentBars(df, genes=unique(genes))
```

```{r heatmap, fig.height=7, fig.width=7}
dreamlet::plotHeatmap(df, genes=unique(genes))
```




```{r Dx.summary, eval=FALSE}
df = unique(data.frame(Dx = pb$dx, Donor = pb$SubID))

sort(table(df$dx), decreasing=TRUE) %>% kbl() %>% kable_styling(full_width=FALSE)
```

```{r voom}
# Normalize and apply voom
form = ~ 1
res.proc = processAssays( pb, form, 
  BPPARAM = SnowParam(6))
  
res.proc
```

```{r voom.plot, fig.height=10, fig.width=10, cache=TRUE}
plotVoom( res.proc ) 
```


```{r plot.sex} 
df_gene =  AnnotationDbi::select(org.Hs.eg.db, keys=c("XIST", "UTY"), keytype="SYMBOL", columns="ENSEMBL")

df_sex = lapply( names(res.proc), function(CT){
  geneExpr = res.proc[[CT]]
  df = NULL
  # if( sum(df_gene$ENSEMBL %in% rownames(geneExpr)) == 2){  
  #   df = data.frame(cellType = CT, IDS = colnames(geneExpr$E), XIST = geneExpr$E[df_gene$ENSEMBL[1],], UTY = geneExpr$E[df_gene$ENSEMBL[2],])
  # }
  if( sum(df_gene$SYMBOL %in% rownames(geneExpr)) == 2){
    df = data.frame(cellType = CT, IDS = colnames(geneExpr$E), XIST = geneExpr$E['XIST',], UTY = geneExpr$E['UTY',])
  }
  df
}) 
df_sex = do.call(rbind, df_sex)

df_sex = merge(df_sex, colData(pb)[,"Sex",drop=FALSE], by.x="IDS", by.y="row.names")

ggplot(as.data.frame(df_sex), aes(XIST, UTY, color=Sex)) + geom_point() + theme_classic() + facet_wrap(~cellType) + theme(aspect.ratio=1)
```

Sex expression based on XIST and UTY
```{r sex.combine, eval=FALSE}
# get total reads
totalReads = DelayedMatrixStats::colSums2(assay(sceCombine, "X"))
df_totalReads = data.frame(totalReads, Channel=sceCombine$Channel) %>%
  group_by(Channel) %>%
  summarise(LibSize = sum(totalReads))

rm(totalReads) 
gc()

# sum across cells for each Donor
counts = assay(sceCombine[df_gene$ENSEMBL,], "X")

setAutoBlockSize(1e8)
grid = colAutoGrid(counts, ncol=100000)

df = dreamlet:::colsum_fast(counts, droplevels(sceCombine$Channel), grid=grid)

# merge expression with metadata
df_meta_uniq = unique(colData(sceCombine)[,c("Channel", "Sex")])
df_meta_uniq = merge(df_meta_uniq, df_totalReads, by="Channel")

df_sex = merge(df_meta_uniq, as.matrix(t(df)), by.x="Channel", by.y="row.names")

rm(df, df_meta_uniq)
gc()

colnames(df_sex)[colnames(df_sex) == df_gene$ENSEMBL[1]] = df_gene$SYMBOL[1] 
colnames(df_sex)[colnames(df_sex) == df_gene$ENSEMBL[2]] = df_gene$SYMBOL[2]

df_sex2 = df_sex %>% 
          group_by(Channel) %>%
          summarise(Sex = unique(Sex),
                    UTY = log2(sum(UTY) + 0.25) - log2(sum(LibSize)) + log2(1e6),
                    XIST= log2(sum(XIST) + 0.25) - log2(sum(LibSize)) + log2(1e6))

ggplot(df_sex2, aes(XIST, UTY, color=Sex)) + geom_point() + theme_classic() + scale_color_manual(values=c("red", "blue", "green"))
```





```{r varPart}
form = ~ (1|SubID) + (1|batch) + (1|Sex) + scale(Age) + (1|AD)

res.vp = fitVarPart(res.proc, form, BPPARAM = SnowParam(6))
```

```{r vp.plot, fig.height=15, fig.width=10}
colnames(res.vp)[colnames(res.vp) == "scale.Age."] = 'Age'
colnames(res.vp)[colnames(res.vp) == "batch"] = 'Batch'
colnames(res.vp)[colnames(res.vp) == "SubID"] = 'Subject'

plotVarPart( sortCols(res.vp), label.angle=45 )
```

```{r plotPercentBars, fig.height=6, fig.width=10}
# Show variance fractions at the gene-level for each cell type
genes = c("KCNMA1", 'RASGRF2', "CECR2", "PTPRG", 'NPNT', 'LINC01844', 'SPRY4-AS1',"XIST", 'C1D')

# res.vp %>%
#  as_tibble %>%
#  filter(assay=="Microglia") %>%
#  summarize(Age = gene[which.max(Age)],
#    Subject = gene[which.max(Subject)],
#    Batch = gene[which.max(Batch)],
#    Residuals = gene[which.max(Residuals)])

# res.vp %>%
#  as_tibble %>%
#  filter(assay=="Microglia") %>%
#  summarize(gene[order(Batch, decreasing=TRUE)][1:5])


df.vp.sub = sortCols(res.vp)[(res.vp$gene %in% genes) &(res.vp$assay=="Microglia"),]
df.vp.sub$gene = factor(df.vp.sub$gene, rev(genes))
df.vp.sub = df.vp.sub[order(df.vp.sub$gene, decreasing=TRUE),]

plotPercentBars(df.vp.sub) + theme(aspect.ratio=1)
```

# Correlation with and between donors
```{r within_btw, fig.height=10, fig.width=10}
source("/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/common_analyses.R")
 
eval_within_across_donor( res.proc, "SubID" )
```

# dreamlet
```{r dreamlet}
form = ~ (1|SubID) + (1|batch) + (1|Sex) + scale(Age) + AD

fit = dreamlet( res.proc, form, BPPARAM = SnowParam(6))
```

## Volcano
```{r volcano, fig.height=10, fig.width=10}
plotVolcano(fit, "AD1")
```

## Highlight genes
```{r genehatmap, fig.height=4, fig.width=6, cache=FALSE}
plotGeneHeatmap( fit, coef="AD1", genes=genes)
```

```{r plotForest, fig.height=6, fig.width=12, cache=FALSE}
figList = lapply(genes, function(g){
  plotForest( fit, coef = 'AD1', gene = g) +
    theme(aspect.ratio=1, legend.position="none")
})
plot_grid(plotlist=figList, ncol=4)
```


### Summarize differential expression 
```{r plot.pi, fig.height=5, fig.width=12}
# Summarize differential expression for each coef and assay
df = fit %>%
  topTable(coef='AD1', number=Inf) %>%
    as_tibble %>% 
    group_by(assay) %>% 
    summarise( 
      nGenes = length(adj.P.Val), 
      nDE = sum(adj.P.Val < 0.05),
      pi1 = 1 - qvalue(P.Value)$pi0) 

ymax = 1.05*max(df$nGenes)
fig1 = ggplot(df, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes expressed") +
    ylab("Cell type") 

ymax = max(1.05*max(df$nDE), 100)
fig2 = ggplot(df, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none", axis.text.y=element_blank()) +
    scale_x_continuous(limits=c(0,ymax), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab('')

fig3 = ggplot(df, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none", axis.text.y=element_blank()) +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1]))+
    ylab('')

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=1, axis="tblr", align="hv")
```    


### Interpreting variance partitioning analysis
Observe that fraction of variation across batches correlates with GC content of the corresponding gene.  Also genes with higher counts per million show high variation across donor since more counts corresponds to a reduction in 'shot noise' due to low counts.

```{r gc.content, fig.height=5, fig.width=12}
source("/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/common_analyses.R")

# read GC content
file = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/misc/GRCh38.104_gc_content.tsv.gz"
df_GC = read.table(file, header=TRUE)

df_vp = merge(as.data.frame(res.vp), df_GC, by.x="gene", by.y="SYMBOL")

fig1 = df_vp %>%
  group_by(assay) %>%
  summarize(cor.se(Batch, GC, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between batch % and GC content") +
    coord_flip() 

tab = topTable(fit, coef='AD1', number=Inf)

df_vp = merge(as.data.frame(res.vp), tab, by.x=c("assay", "gene"), by.y=c("assay", "ID") )

fig2 = df_vp %>% 
  as_tibble %>%
  group_by(assay) %>%
  summarize(cor.se(Subject, AveExpr, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank(), 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between Donor % and CPM") +
    scale_y_continuous(limits=c(0, .7), expand=c(0,0)) +
    coord_flip() 

plot_grid(fig1, fig2, labels=LETTERS[1:2], nrow=1, align="hv", axis="tblr")    
```

## Gene set analysis
```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="SYMBOL")

# zenith analysis
res.gsa = zenith_gsa(fit, go.gs, coefs="AD1", n_genes_min=30)
```

```{r zenith.heatmap.1, fig.width=10, fig.height=18, cache=FALSE} 
plotZenithResults(res.gsa, 5, 2) + theme(legend.position="bottom") 
```


# Gene set analysis using zenith
```{r zenith2}
go.gs.en = getGenesets( org = "hsa", 
                  db = "enrichr", 
                  lib = "GO_Biological_Process_2021",
                  gene.id.type = "SYMBOL", 
                  return.type = "GeneSetCollection")

res.zenith2 = zenith_gsa( fit, go.gs.en, 
              coef = 'AD1', 
              n_genes_min=20)
```

```{r plot.zenith2, fig.width=15, fig.height=16}
res.zenith2$assay = factor(res.zenith2$assay, names(fit))
plotZenithResults(res.zenith2, 3, 1)
```

# Mashr Analysis
```{r mashr, eval=FALSE}
# run mashr model to borrow information across genes and
# cell types in estimating coefficients' posterior distribution
res_mash = run_mash(fit, coef='AD1')
```




# Session Info
<details>
```{r sessioninfo, cache=FALSE}
sessionInfo()
```
</details>



```{r exit2, cache=FALSE}
knitr::knit_exit()
```












