---
title: "Analysis of single cell RNA-seq in autism"
subtitle: '[Velmeshev, et al. 2019](https://www.science.org/doi/10.1126/science.aav8130)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---



<!---

cd /sc/arion/projects/CommonMind/hoffman/dreamlet_analysis/Velmeshev_2019
ml git
R

system("git pull"); rmarkdown::render("Velmeshev_2019.Rmd")



https://hoffmg01.hpc.mssm.edu/dreamlet_analysis/Velmeshev_2019/Velmeshev_2019.html

--->


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  package.startup.message = FALSE,
  cache.lazy = FALSE)
```

# Load packages
```{r pkgs, cache=FALSE}
library(zellkonverter)
library(SingleCellExperiment)
library(dreamlet)
library(ggplot2)
library(tidyverse)
library(aplot)
library(ggtree)
library(scattermore)
library(zenith)
library(crumblr)
library(GSEABase)
library(qvalue)
library(ggcorrplot)
library(BiocParallel)
library(cowplot)
library(DelayedArray)
setAutoBlockSize(1e9)

trimGenesets = function(x){
  x %>%
    gsub("^(ER\\d+)_", "", .) %>% 
    gsub("_\\(GO.*\\)$", "", .) %>% 
    gsub("_", " ", .)
}
```

# Load data
```{r load.data}
# read H5AD files
file = "/sc/arion/projects/CommonMind/hoffman/crumblr_analysis/Velmeshev_2019/data/Velmeshev_2019.h5ad"
sce = readH5AD(file, use_hdf5=TRUE)

# keep only features with unique gene symbols
keep = ! duplicated(rowData(sce)$SYMBOL)
sce = sce[keep,]
rownames(sce) = rowData(sce)$SYMBOL

pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cluster",  
    sample_id = "sample", 
    BPPARAM = SnowParam(6, progressbar=TRUE))
```

## tNSE
```{r tsne}
reducedDim(sce, "tSNE") %>%
  as_tibble %>%
  cbind( cluster = sce$cluster) %>% 
  ggplot(aes(tsne1,tsne2, color=cluster)) +
    geom_scattermore() +
    theme_void() +
    theme(aspect.ratio=1)
```
    

```{r cca}
form = ~ individual + region + age + sex + diagnosis + Capbatch + Seqbatch + post.mortem.interval..hours. + RNA.Integrity.Number

C = canCorPairs(form, colData(pb) )
ggcorrplot(C, hc.order = TRUE)
```



# Process assays
```{r dreamlet}
form = ~ (1|individual) + (1|region) + age + (1|sex) + diagnosis + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
res.proc = processAssays( pb, form, BPPARAM = SnowParam(6))
```
## Mean variance trend
```{r voom, fig.height=12}
plotVoom( res.proc)
```


# Variance partitioning analysis
```{r vp}
form = ~ (1|individual) + (1|region) + age + (1|sex) + (1|diagnosis) + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
vp.lst = fitVarPart( res.proc, form, BPPARAM = SnowParam(6))
```


```{r plotVarPart, fig.height=12, fig.width=9}
plotVarPart(sortCols(vp.lst), label.angle=60) 
```


```{r vp2}
form = ~ (1|individual) + (1|region) + age + (1|sex) + (1|diagnosis) + post.mortem.interval..hours. + RNA.Integrity.Number
vp.lst2 = fitVarPart( res.proc, form, BPPARAM = SnowParam(6))
```


```{r plotVarPart2, fig.height=12, fig.width=9}
plotVarPart( sortCols(vp.lst2), label.angle=90, ncol=4) 
```


```{r plotVarPart3, fig.height=4, fig.width=4}
vp = vp.lst2[vp.lst2$assay == "L2/3",]
plotVarPart( sortCols(vp), label.angle=90) 
```

## Interpreting variance partitioning analysis
Observe that fraction of variation across batches correlates with GC content of the corresponding gene. Also genes with higher counts per million show high variation across donor since more counts corresponds to a reduction in ‘shot noise’ due to low counts.

```{r gc, fig.height=5, fig.width=12}
# read GC content
path = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/"
file = paste0(path, '/misc/GRCh38.104_gc_content.tsv.gz')
df_GC = read.table(file, header=TRUE)

df_vp = merge(as.data.frame(vp.lst), df_GC, by.x="gene", by.y="SYMBOL")

# Compute correlation and standard error for Pearson and Spearman
cor.se = function(x,y, method = c("pearson", "kendall", "spearman"),...){

  method = match.arg(method)

  if( method == "pearson"){
    df <- cor.test(x,y, method=method,...) %>%
      tidy %>%
      mutate(se = sqrt((1 - estimate^2)/parameter))
    res = data.frame(rho = df$estimate, rho.se = df$se)
  }else if(method == "spearman"){
    # https://stats.stackexchange.com/questions/18887/how-to-calculate-a-confidence-interval-for-spearmans-rank-correlation
    rho <- cor(x,y, method=method,...)
    n <- sum(complete.cases(x, y))
    rho.se <- sqrt((1+rho^2/2)/(n-3))

    res <- data.frame(rho = rho, rho.se = rho.se)
  }

  return(res) 
}

fig1 = df_vp %>%
  group_by(assay) %>%
  mutate(batch = Capbatch + Seqbatch) %>%
  summarize(cor.se(batch, GC, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between batch % and GC content") +
    coord_flip() 

tab = topTable(res.dl, 'diagnosisASD', number=Inf)

df_vp = merge(as.data.frame(vp.lst), tab, by.x=c("assay", "gene"), by.y=c("assay", "ID") )

fig2 = df_vp %>% 
  as_tibble %>%
  group_by(assay) %>%
  summarize(cor.se(individual, AveExpr, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank(), 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between Donor % and CPM") +
    coord_flip() 

plot_grid(fig1, fig2, labels=LETTERS[1:2], nrow=1, align="hv", axis="tblr") 
```
# Differential expression
```{r run.de}
form = ~ (1|individual) + (1|region) + age + (1|sex) + diagnosis + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
res.dl = dreamlet( res.proc, form, BPPARAM = SnowParam(6) )
```


```{r plotVolcano, fig.height=12, fig.width=9}
plotVolcano( res.dl, coef = 'diagnosisASD', ncol=4 )
```

## Summarize differential expression 
```{r plot.pi, fig.height=5, fig.width=17}
tab = topTable(res.dl, coef = 'diagnosisASD', number=Inf) %>%
  as_tibble %>% 
  group_by(assay) %>%
  summarize( nDE = sum(adj.P.Val < 0.05), 
            pi1 = 1 - pi0est(P.Value)$pi0,
            nGenes = length(adj.P.Val)) %>%
  mutate(assay = factor(assay, assayNames(pb)))

fig1 = ggplot(tab, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,NA), expand=c(0,10)) +
    xlab("# genes expressed") +
    ylab("Cell type") 

fig2 = ggplot(tab, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, axis.text.y=element_blank(), axis.ticks.y = element_blank(), legend.position="none") +
    scale_x_continuous(limits=c(0,1000), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab("Cell type")

fig3 = ggplot(tab, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, axis.text.y=element_blank(), axis.ticks.y = element_blank(), legend.position="none") +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1])) +
    ylab("Cell type")

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=1, axis="tblr", align="hv")
```    


# Gene set analysis
```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="SYMBOL")
    
# Run zenith gene set analysis on result of dreamlet
res_zenith = zenith_gsa(res.dl, coef = 'diagnosisASD', go.gs)
```


```{r plotZenithResults, fig.height=12, fig.width=10}
plotZenithResults(res_zenith, 5, 1) 
```



## Gene set with FDR < 5%
```{r plotZenithResults.FDR, fig.height=12, fig.width=10}
# get genesets with FDR < 5%
gs = unique(res_zenith$Geneset[res_zenith$FDR < 0.05])

# keep only results of these genesets 
df = res_zenith[res_zenith$Geneset %in% gs,]

# plot results, but with no limit based on the highest/lowest t-statistic
plotZenithResults(df, Inf, Inf)
```



### Highlight specific genesets
```{r panels3, fig.height=3, fig.width=8}
CT = "L2/3"
gs1 = "GO0099172: presynapse organization"
plotGeneHeatmap( res.dl, coef="diagnosisASD", genes=geneIds(go.gs[[gs1]]), assays=CT, transpose=TRUE) + 
  ggtitle(trimGenesets(gs1)) + 
  theme(legend.position = "bottom", 
    axis.text.x=element_text(size=6, angle=90),
    axis.text.y=element_text(size=6)) + 
  xlab('') + ylab('') 
```

