---
title: "Analysis of single cell RNA-seq in autism"
subtitle: '[Velmeshev, et al. 2019](https://www.science.org/doi/10.1126/science.aav8130)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---



<!---

cd /sc/arion/projects/CommonMind/hoffman/dreamlet_analysis/Velmeshev_2019
ml git
R

system("git pull"); rmarkdown::render("Velmeshev_2019.Rmd")



https://hoffmg01.hpc.mssm.edu/dreamlet_analysis/Velmeshev_2019/Velmeshev_2019.html

--->


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  package.startup.message = FALSE,
  cache.lazy = FALSE)
```

# Load packages
```{r pkgs, cache=FALSE}
library(zellkonverter)
library(SingleCellExperiment)
library(dreamlet)
library(ggplot2)
library(tidyverse)
library(aplot)
library(ggtree)
library(scattermore)
library(zenith)
library(crumblr)
library(GSEABase)
library(ggcorrplot)
library(BiocParallel)

library(DelayedArray)
setAutoBlockSize(1e9)

trimGenesets = function(x){
  x %>%
    gsub("^(ER\\d+)_", "", .) %>% 
    gsub("_\\(GO.*\\)$", "", .) %>% 
    gsub("_", " ", .)
}
```

# Load data
```{r load.data}
# read H5AD files
file = "/sc/arion/projects/CommonMind/hoffman/crumblr_analysis/Velmeshev_2019/data/Velmeshev_2019.h5ad"
sce = readH5AD(file, use_hdf5=TRUE)

# keep only features with unique gene symbols
keep = ! duplicated(rowData(sce)$SYMBOL)
sce = sce[keep,]
rownames(sce) = rowData(sce)$SYMBOL

pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cluster",  
    sample_id = "sample", 
    BPPARAM = SnowParam(6, progressbar=TRUE))
```

## tNSE
```{r tsne}
reducedDim(sce, "tSNE") %>%
  as_tibble %>%
  cbind( cluster = sce$cluster) %>% 
  ggplot(aes(tsne1,tsne2, color=cluster)) +
    geom_scattermore() +
    theme_void() +
    theme(aspect.ratio=1)
```
    

```{r cca}
form = ~ individual + region + age + sex + diagnosis + Capbatch + Seqbatch + post.mortem.interval..hours. + RNA.Integrity.Number

C = canCorPairs(form, colData(pb) )
ggcorrplot(C, hc.order = TRUE)
```



# dreamlet
```{r dreamlet}
form = ~ (1|individual) + (1|region) + age + (1|sex) + diagnosis + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
res.proc = processAssays( pb, form, BPPARAM = SnowParam(6))
```


```{r voom, fig.height=12}
plotVoom( res.proc)
```

```{r vp}
form = ~ (1|individual) + (1|region) + age + (1|sex) + (1|diagnosis) + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
vp.lst = fitVarPart( res.proc, form, BPPARAM = SnowParam(6))
```


```{r plotVarPart, fig.height=12, fig.width=9}
plotVarPart(sortCols(vp.lst), label.angle=60) 
```


```{r vp2}
form = ~ (1|individual) + (1|region) + age + (1|sex) + (1|diagnosis) + post.mortem.interval..hours. + RNA.Integrity.Number
vp.lst2 = fitVarPart( res.proc, form, BPPARAM = SnowParam(6))
```


```{r plotVarPart2, fig.height=12, fig.width=9}
plotVarPart( sortCols(vp.lst2), label.angle=90, ncol=4) 
```


```{r plotVarPart3, fig.height=4, fig.width=4}
vp = vp.lst2[vp.lst2$assay == "L2/3",]
plotVarPart( sortCols(vp), label.angle=90) 
```


```{r run.de}
form = ~ (1|individual) + (1|region) + age + (1|sex) + diagnosis + (1|Capbatch) + (1|Seqbatch) + post.mortem.interval..hours. + RNA.Integrity.Number
res.dl = dreamlet( res.proc, form, BPPARAM = SnowParam(6) )
```


```{r plotVolcano, fig.height=12, fig.width=9}
plotVolcano( res.dl, coef = 'diagnosisASD', ncol=4 )
```

```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology(to="SYMBOL")
    
# Run zenith gene set analysis on result of dreamlet
res_zenith = zenith_gsa(res.dl, coef = 'diagnosisASD', go.gs)
```


```{r plotZenithResults, fig.height=12, fig.width=10}
plotZenithResults(res_zenith, 5, 1) 
```



## Gene set with FDR < 5%
```{r plotZenithResults.FDR, fig.height=12, fig.width=10}
# get genesets with FDR < 5%
gs = unique(res_zenith$Geneset[res_zenith$FDR < 0.05])

# keep only results of these genesets 
df = res_zenith[res_zenith$Geneset %in% gs,]

# plot results, but with no limit based on the highest/lowest t-statistic
plotZenithResults(df, Inf, Inf)
```



### Highlight specific genesets
```{r panels3, fig.height=3, fig.width=8}
CT = "L2/3"
gs1 = "GO0099172: presynapse organization"
plotGeneHeatmap( res.dl, coef="diagnosisASD", genes=geneIds(go.gs[[gs1]]), assays=CT, transpose=TRUE) + 
  ggtitle(trimGenesets(gs1)) + 
  theme(legend.position = "bottom", 
    axis.text.x=element_text(size=6, angle=90),
    axis.text.y=element_text(size=6)) + 
  xlab('') + ylab('') 
```

