---
title: "Multimodally profiling memory T cells from a tuberculosis cohort identifies cell state associations with demographics, environment and disease (GSE158769)"
subtitle: '[Nathan, et al. Nat Immunol  (2021)](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158769)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
    toc: true
    smart: false
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```

Perform dreamlet analysis of single cell data assaying T cell populations in donors infected with tuberculosis.

# Loading
## Load libraries
```{r load.packages, cache=FALSE}
library(GSEABase)
library(SingleCellExperiment)
library(zellkonverter)
library(dreamlet)
library(scater)
library(ggplot2)
library(zenith)
library(scattermore)
library(tidyverse)
library(knitr)
library(broom)
library(qvalue)
library(cowplot)
library(RColorBrewer)
library(kableExtra)
library(EnrichmentBrowser)
library(DelayedArray)
```

# Load data
```{r load.data}
# read H5AD file
path = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Nathan_NatImm_2021/"
file = paste0(path, "/Nathan_NatImm_2021.h5ad")
sce = readH5AD(file, use_hdf5=TRUE)   

# get only single-cell RNA-seq
# rest is CITE-seq proteins
only_rna = grep("prot", rownames(sce), invert=TRUE)

# create variable that distinguishes cells from the same donor
# but different batches
sce$donor_batch = with(colData(sce), paste(donor, batch))
sce$batch = factor(sce$batch)
sce$season = factor(sce$season)
sce$TB_status = factor(sce$TB_status, c("CONTROL", "CASE"))

# sort cell clusters by name
sce = sce[only_rna,sce$cluster_name != "NA"]
colData(sce) = droplevels(colData(sce))
tab = colData(sce) %>%
  as_tibble %>%
  group_by(cluster_ids, cluster_name) %>%
  filter(row_number()==1) %>%
  summarise(cluster_ids, cluster_name) %>%
  arrange(as.numeric(gsub("C-", '', cluster_ids)))

sce$cluster_name = factor(sce$cluster_name, unique(tab$cluster_name))
sce$cluster_ids = factor(sce$cluster_ids, unique(tab$cluster_ids))

# create pseudobulk 
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cluster_name",  
    sample_id = "donor_batch")

# Gene expressed genes for each cell type
geneList = getExprGeneNames(pb, min.samples = 10)

# Compute weights
W.list <- pbWeights( sce, 
                    cluster_id = "cluster_name", 
                    sample_id = "donor_batch",
                    details = TRUE, 
                    method = "delta", 
                    geneList = geneList)
```


## plots
```{r vif}
# extract results
df <- lapply(names(W.list), function(x){
  attr(W.list[[x]], "details") 
})
df <- bind_rows(df) 

ggplot(df, aes(assay, vif)) +
  geom_violin(scale="width") +  
  geom_boxplot(width=0.1, fill="grey70") +
  theme_classic() +
  theme(aspect.ratio=1) +
  scale_y_log10() + 
  geom_hline(yintercept=1, color="red", linetype="dashed") +
  ylab("Variance Inflation Factor") +
  coord_flip()

ggplot(df, aes(assay, sigSq.mle)) +
  geom_violin(scale="width") +  
  geom_boxplot(width=0.1, fill="grey70") +
  theme_classic() +
  theme(aspect.ratio=1) +
  scale_y_log10() + 
  ylab("sigSq.mle") +
  coord_flip()

ggplot(df, aes(assay, sigSq.hat)) +
  geom_violin(scale="width") +  
  geom_boxplot(width=0.1, fill="grey70") +
  theme_classic() +
  theme(aspect.ratio=1) +
  scale_y_log10() + 
  ylab("sigSq.hat") +
  coord_flip()

df %>%
  select(Gene, assay, sigSq.hat, sigSq.mle) %>%
  pivot_longer(cols=c("sigSq.mle", "sigSq.hat")) %>%
  ggplot(aes(assay, value, color=name)) +
    # geom_violin(scale="width", color=NA) +  
    geom_boxplot(width=0.1, fill="grey70") +
    theme_classic() +
    theme(aspect.ratio=2) +
    scale_y_log10() + 
    coord_flip()
```

## Plot UMAP
```{r UMAP}
colData(sce) %>%  
  data.frame %>%
  ggplot(aes(UMAP_1, UMAP_2, color=cluster_name)) +
    geom_scattermore() +
    theme_classic() +
    theme(aspect.ratio=1)
```

## Show identifiers for each cell cluster
```{r cell.ids}
colData(sce) %>%
  as_tibble %>%
  group_by(cluster_ids, cluster_name) %>%
  filter(row_number()==1) %>%
  summarise(cluster_ids, cluster_name) %>%
  arrange(as.numeric(gsub("C-", '', cluster_ids))) %>% 
  kbl(row.names=FALSE) %>% 
  kable_classic(full_width = FALSE) 
```

## Properties of dataset
```{r properties}
df_properties = list()

# number of batches
df_properties$`# batches` = nlevels(sce$batch)

# number of donors
df_properties$`# donors` = nlevels(sce$donor)

df_properties$`# donors with technical replicates` = sum(table(pb$donor) > 1)

# number of donors
df = unique(data.frame(donor = pb$donor, TB_status = pb$TB_status))
df_properties$`# Cases` = sum(df$TB_status == "CASE")
df_properties$`# Controls` = sum(df$TB_status == "CONTROL")

df_properties$`# Cells` = ncol(sce)

df_properties = unlist(df_properties)
data.frame(Property = names(df_properties), counts = df_properties) %>% 
  kbl(row.names=FALSE) %>% 
  kable_classic(full_width = FALSE) 
```


## Relationship between variables
Use canonical correlation analysis to show association between variables
```{r cca, fig.height=6, fig.width=6}
form = ~ TB_status + sex + age + donor + batch + season + prop_NAT
C = canCorPairs(form, colData(pb))
 
plotCorrMatrix(C)
```

## Process pseudobulk data to estimate precision weights
```{r voom}
form = ~ TB_status + (1|donor) + (1|batch) + (1|season) + (1|sex) + scale(age) + I(scale(age^2)) + scale(prop_NAT)
res.proc = processAssays( pb, form, 
              min.samples = 10,
              BPPARAM = SnowParam(12, progressbar=TRUE), 
              weightsList = W.list )
```

### Show details of processing
```{r details}
details(res.proc)
```

### Show voom-style mean-variance trends
```{r plot.voom, fig.width=8, fig.height=16}
plotVoom( res.proc, ncol=4, alpha=.1)  
```

# Variance partitioning analysis
Note that variation across donors here indicates technical variation since the multiple samples from the same donor are technical replicates.
```{r run.vp}
form = ~ (1|TB_status) + (1|donor) + (1|batch) + (1|season) + (1|sex) + scale(age) + I(scale(age)^2) + scale(prop_NAT)
vp.lst = fitVarPart( res.proc, form, 
  BPPARAM=SnowParam(12, progressbar=TRUE)) 
```

```{r plot.vp, fig.width=8, fig.height=16}
plotVarPart(vp.lst, label.angle=60, ncol=4)     
```

# dreamlet analysis
```{r dreamlet}
form = ~ TB_status + (1|donor) + (1|batch) + (1|season) + (1|sex) + scale(age) + I(scale(age)^2) + scale(prop_NAT)
res.dl = dreamlet( res.proc, form, 
  BPPARAM=SnowParam(12, progressbar=TRUE))
```

### Volcano plot for each cell type
```{r volcano, fig.width=9, fig.height=15}
plotVolcano( res.dl, coef = 'TB_statusCASE', ncol=4) 
```


### Summarize differential expression 
```{r plot.pi, fig.height=5, fig.width=17}
tab = topTable(res.dl, coef = 'TB_statusCASE', number=Inf) %>%
  as_tibble %>% 
  group_by(assay) %>%
  summarize( nDE = sum(adj.P.Val < 0.05), 
            pi1 = 1 - pi0est(P.Value)$pi0,
            nGenes = length(adj.P.Val)) %>%
  mutate(assay = factor(assay, assayNames(pb)))

fig1 = ggplot(tab, aes(nGenes, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_x_continuous(limits=c(0,NA), expand=c(0,10)) +
    xlab("# genes expressed") +
    ylab("Cell type") 

fig2 = ggplot(tab, aes(nDE, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, axis.text.y=element_blank(), axis.ticks.y = element_blank(), legend.position="none") +
    scale_x_continuous(limits=c(0,1000), expand=c(0,0)) +
    xlab("# genes with FDR < 5%") +
    ylab("Cell type")

fig3 = ggplot(tab, aes(pi1, assay, fill=assay)) + 
    geom_bar(stat="identity") + 
    theme_classic() +
    theme(aspect.ratio=1, axis.text.y=element_blank(), axis.ticks.y = element_blank(), legend.position="none") +
    scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
    xlab(bquote(pi[1])) +
    ylab("Cell type")

plot_grid(fig1, fig2, fig3, labels=LETTERS[1:3], nrow=1, axis="tblr", align="hv")
```    




### Interpreting variance partitioning analysis
Observe that fraction of variation across batches correlates with GC content of the corresponding gene.  Also genes with higher counts per million show high variation across donor since more counts corresponds to a reduction in 'shot noise' due to low counts.

```{r gc.content, fig.height=5, fig.width=12}
# read GC content
file = paste0(path, '/../misc/GRCh38.104_gc_content.tsv.gz')
df_GC = read.table(file, header=TRUE)

df_vp = merge(as.data.frame(vp.lst), df_GC, by.x="gene", by.y="SYMBOL")

# Compute correlation and standard error for Pearson and Spearman
cor.se = function(x,y, method = c("pearson", "kendall", "spearman"),...){

  method = match.arg(method)

  if( method == "pearson"){
    df <- cor.test(x,y, method=method,...) %>%
      tidy %>%
      mutate(se = sqrt((1 - estimate^2)/parameter))
    res = data.frame(rho = df$estimate, rho.se = df$se)
  }else if(method == "spearman"){
    # https://stats.stackexchange.com/questions/18887/how-to-calculate-a-confidence-interval-for-spearmans-rank-correlation
    rho <- cor(x,y, method=method,...)
    n <- sum(complete.cases(x, y))
    rho.se <- sqrt((1+rho^2/2)/(n-3))

    res <- data.frame(rho = rho, rho.se = rho.se)
  }

  return(res) 
}

fig1 = df_vp %>%
  group_by(assay) %>%
  summarize(cor.se(batch, GC, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between batch % and GC content") +
    coord_flip() 

tab = topTable(res.dl, 'TB_statusCASE', number=Inf)

file = "./topTable_Nathan_NatImm_2021.tsv"
write.table(tab, file, quote=FALSE, sep='\t', row.names=FALSE)


df_vp = merge(as.data.frame(vp.lst), tab, by.x=c("assay", "gene"), by.y=c("assay", "ID") )

fig2 = df_vp %>% 
  as_tibble %>%
  group_by(assay) %>%
  summarize(cor.se(donor, AveExpr, method="spearman")) %>%
  ggplot(aes(assay, rho, fill=assay)) +
    geom_bar(stat="identity") + 
    geom_errorbar(aes(ymin= rho - 1.96*rho.se, ymax= rho + 1.96*rho.se), width=0) +
    theme_classic() +
    theme(aspect.ratio=1, 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank(), 
      legend.position="none") +
    xlab("Cell type") +
    ylab("Spearman correlation between Donor % and CPM") +
    coord_flip() 

plot_grid(fig1, fig2, labels=LETTERS[1:2], nrow=1, align="hv", axis="tblr")    
```


# Gene set analysis using zenith
```{r zenith}
go.gs = getGenesets( org = "hsa", 
                  db = "enrichr", 
                  lib = "GO_Biological_Process_2021",
                  gene.id.type = "SYMBOL", 
                  return.type = "GeneSetCollection")

res.zenith = zenith_gsa( res.dl, go.gs, 
              coef = 'TB_statusCASE', 
              n_genes_min=20)
```

```{r plot.zenith, fig.width=15, fig.height=16}
res.zenith$assay = factor(res.zenith$assay, names(res.dl))
plotZenithResults(res.zenith, 3, 1)
```


### Heatmap of results
```{r plot.zenith.subset, fig.height=4, fig.width=12}
ids = c("GO0006613", "GO0043043", "GO0050852", "GO1902750", "GO0071346", "GO0038094")

idx = grep( paste0(ids, collapse="|"), res.zenith$Geneset)
df2 = res.zenith[idx,]

# Clean up  geneset names
trimGenesets = function(x){
  x %>%
    gsub("^(ER\\d+)_", "", .) %>% 
    gsub("_\\(GO.*\\)$", "", .) %>% 
    gsub("_", " ", .)
}

df2$Geneset = trimGenesets(df2$Geneset)

plotZenithResults(df2, 3, 1, transpose=FALSE) + 
  theme(axis.text.x=element_text(size=10, angle = 30),
    axis.text.y=element_text(size=10), 
    legend.position="right", 
    legend.key.size = unit(.3, 'cm') ) 
```




### Select genes with high variation explained by each variable 
```{r plotPercent}
CT = "CD4/8+ PD-1+TIGIT+"
ngenes = 1
genes = vp.lst %>%
  as_tibble %>%
  filter(assay == CT) %>%
  summarize(gene[order(batch, decreasing=TRUE)[1:ngenes]],
            gene[order(donor, decreasing=TRUE)[1:ngenes]],
            gene[order(season, decreasing=TRUE)[1:ngenes]],
            gene[order(sex, decreasing=TRUE)[1:ngenes]],
            gene[order(TB_status, decreasing=TRUE)[1:ngenes]],
            gene[order(age + I.age.2., decreasing=TRUE)[1:ngenes]],
            gene[order(prop_NAT, decreasing=TRUE)[1:ngenes]],
            gene[order(Residuals, decreasing=TRUE)[1:ngenes]]) %>%
  unlist %>% 
  array

genes = c("HIST1H1E", "EIF1", "CXCR4", "RPS4X", "LTB", "TSPO",   "RPS2", "NDFIP1")

col = c(brewer.pal(n = 8, name = "Set1"), "grey85")

fig.percent = plotPercentBars(vp.lst, assays = CT, genes = genes, col = col)  
```

### Combine figure Panels
```{r panels, fig.height=4, fig.width=12}
fig.umap = colData(sce) %>%
  data.frame %>% 
  ggplot(aes(UMAP_1, UMAP_2, color=ifelse(cluster_name==CT, CT, ''))) +
    geom_scattermore() +
    theme_classic() +
    scale_color_manual(name='', values=c("grey", "orange")) +
    theme(aspect.ratio=1, 
      axis.text.x=element_blank(), 
      axis.text.y=element_blank(), 
      axis.ticks=element_blank(), 
      legend.position="bottom") +
    xlab('UMAP 1') +
    ylab('UMAP 2') 

thm = theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), aspect.ratio=1)

fig.voom = plotVoom( res.proc[[CT]], alpha=.8) + theme_classic() + thm
fig.vp = plotVarPart(vp.lst, assays = CT, label.angle=40, col=col) +  
  thm +    
  theme(axis.text.x=element_text(size=10))
file = paste0(getwd(), "/vp.pdf")
ggsave(file, fig.vp)

fig.vol = plotVolcano( res.dl[[CT]], coef = 'TB_statusCASE')

fig.pct = fig.percent + theme(aspect.ratio=1, legend.position="bottom")

plot_grid(plotlist = list(fig.umap, fig.voom, fig.vp, fig.pct), nrow=1, labels=LETTERS[1:4], align="hv", axis="t")
```

### Highlight specific genes
```{r panels2, fig.height=6}
data = extractData(res.proc, CT)

fig.strat1 = plotStratify( HIST1H1E ~ batch, data, legend=FALSE, colorBy=NULL) + 
  theme(aspect.ratio=1) +
  ylab(bquote(log[2]~CPM)) + 
  ggtitle("HIST1H1E") 

fig.strat2 = plotStratify( LTB ~ TB_status, data, x.labels=TRUE, sort=FALSE) +
  theme(aspect.ratio=1, legend.position="none", axis.text.x=element_text(angle=0)) +
  scale_fill_manual(values=c("blue", "red")) +
  scale_color_manual(values=c("blue", "red"))+
  ylab(bquote(log[2]~CPM)) + 
  ggtitle("LTB")

plot_grid(fig.strat1, fig.strat2, align="hv", axis="tblr", rel_widths=c(1.5,1), labels=LETTERS[5:6])
```

### Highlight specific genesets
```{r panels3, fig.height=3, fig.width=8}
gs1 = "ER541_cellular_response_to_interferon-gamma_(GO0071346)"
plotGeneHeatmap( res.dl, coef="TB_statusCASE", genes=geneIds(go.gs[[gs1]]), assays=CT, transpose=TRUE) + 
  ggtitle(trimGenesets(gs1)) + 
  theme(legend.position = "bottom",
    axis.text.x=element_text(size=10, angle=30),
    axis.text.y=element_text(size=10)) + 
  xlab('') + ylab('') 
```


# Manuscript text
```{r summary.for.text}
vp.summary_batch = vp.lst %>%
  as_tibble %>%
  filter(assay == CT) %>%
  select(batch) %>%
  summarize(median = median(batch), 
    fraction = sum(batch > .25) / length(batch))

vp.summary_TB = vp.lst %>%
  as_tibble %>%
  filter(assay == CT) %>%
  select(TB_status) %>%
  summarize(median = median(TB_status), 
    fraction = sum(TB_status > .01) / length(TB_status))
```

Variance partitioning analysis for each gene in this cluster estimates the fraction of expression variance attributable to variation across `r nlevels(pb$batch)` technical batches, `r nlevels(pb$donor)` donors and 2 disease states (Figure XC).  While the median fraction of variation explained by batch is `r format(vp.summary_batch$median *100, digits=3)`%, `r format(vp.summary_batch$fraction *100, digits=3)`% of expressed genes have more than 25% of their expression variance attributed to cross-batch variation.  Importantly, TB status explains > 1% of variance for just `r format(vp.summary_TB$fraction *100, digits=3)`% of expressed genes, underscoring the small effect of TB infection following disease resolution. 


Number of tests: `r topTable(res.dl, coef = 'TB_statusCASE', number=Inf) %>% nrow`

# Session info
<details>
```{r sessionInfo}
sessionInfo()
```
</details>

```{r exit123, cache=FALSE, echo=FALSE}
knitr::knit_exit()
```





# Cell type composition

```{r crumblr}
library(crumblr)

# run crumblr transformation
cobj = crumblr(cellCounts(pb))

# variance partitioning analysis
form = ~ (1|TB_status) + (1|donor) + (1|batch) + (1|season) + (1|sex) + age + I(age^2) + prop_NAT
vp_res = fitExtractVarPartModel(cobj, form, colData(pb))
plotVarPart(sortCols(vp_res))

# PCA
E_vst = vst(cobj)
dcmp = prcomp(t(E_vst))

biplot(dcmp)

dcmp$x %>%
  merge(colData(pb), by="row.names") %>%
  as.data.frame %>%
  ggplot(aes(PC1, PC2, color=season)) +
    geom_point() +
    theme_classic() +
    theme(aspect.ratio=1)

# differential abundance analysis
fit = dream(cobj, ~ TB_status + (1|donor) + (1|batch) + (1|season) + (1|sex) + age + I(age^2) + prop_NAT, colData(pb))
fit = eBayes(fit)
topTable(fit, coef='TB_statusCASE', number=Inf) %>%   
  select(logFC, AveExpr, t, P.Value, adj.P.Val) %>% 
  kbl() %>% 
  kable_classic(full_width = FALSE) 
```

















Despite the heterogeneity of T-cell expression profiles, we detect as shared component of response to stimulus

## MSigDB
```{r zenith.ms}
go.ms = get_MSigDB(to="SYMBOL")

res.zenith.ms = zenith_gsa( res.dl, go.ms, 'TB_statusCASE')
```

```{r plot.zenith.ms, fig.width=7, fig.height=16}
plotZenithResults(res.zenith.ms, 3, 1)
```




```{r mashr}
res_mash = run_mash(res.dl, coef='TB_statusCASE')
```


Vd1: C-31

```{r plot.TRBC1}
plotForest(res.dl, "CAST", coef='TB_statusCASE') 
```

```{r plot.TRBC1.mashr}
plotForest(res_mash, "CAST") 
```


```{r mashr.downstream}
# how many gene-by-celltype tests are significant
# i.e.  if a gene is significant in 2 celltypes, it is counted twice
table(get_lfsr(res_mash$model) < 0.05, useNA="ifany")

# how many genes are significant in at least one cell type
table( apply(get_lfsr(res_mash$model), 1, min, na.rm=TRUE) < 0.05)

# how many genes are significant in each cell type
apply(get_lfsr(res_mash$model), 2, function(x) sum(x < 0.05, na.rm=TRUE))

df_gs = zenith_gsa(res_mash, go.gs)
```

```{r mashr.heatmap, fig.width=7, fig.height=10}
# Heatmap of results
plotZenithResults(df_gs, 3, 1)
```

```{r mashr.volcano, fig.width=8, fig.height=16}
# volcano plot based on mashr results
# yaxis uses local false sign rate (lfsr)
plotVolcano(res_mash, ncol=4)
```
















